### DISCUSSION TITLE
Bug in DT 4.7?

### DISCUSSION THREAD

**Discussion by JoergBruenner** [DependencyTrack/dependency-track#2348] [Q&A] [answered]
Hi all,

I'm merging two SBoMs by the CycloneDX CLI Tool. The result isn't shown correct in DT.
The files to merge are:

merge1-1:
{
  "bomFormat": "CycloneDX",
  "specVersion": "1.4",
  "serialNumber": "urn:uuid:3e671687-395b-41f5-a30f-a58921a69b79",
  "version": 1,
  "metadata": {
    "component":     
	{
      "bom-ref": "pkg:maven/org.acme/persistence@3.1.0",
      "type": "library",
      "group": "org.acme",
      "name": "persistence",
      "version": "3.1.0",
      "purl": "pkg:maven/org.acme/persistence@3.1.0"
    }
  },
  "components": [
    {
      "bom-ref": "pkg:maven/org.acme/common-util@3.0.0",
      "type": "library",
      "group": "org.acme",
      "name": "common-util",
      "version": "3.0.0",
      "purl": "pkg:maven/org.acme/common-util@3.0.0"
    }
  ],
  "dependencies": [
    {
      "ref": "pkg:maven/org.acme/persistence@3.1.0",
      "dependsOn": [
        "pkg:maven/org.acme/common-util@3.0.0"
      ]
    },
    {
      "ref": "pkg:maven/org.acme/common-util@3.0.0",
      "dependsOn": []
    }
  ]
}
merge1-2:
{
  "bomFormat": "CycloneDX",
  "specVersion": "1.4",
  "serialNumber": "urn:uuid:3e671687-395b-41f5-a30f-a58921a69b79",
  "version": 1,
  "metadata": {
    "component": 
    {
      "bom-ref": "pkg:maven/org.acme/web-framework@1.0.0",
      "type": "library",
      "group": "org.acme",
      "name": "web-framework",
      "version": "1.0.0",
      "purl": "pkg:maven/org.acme/web-framework@1.0.0"
    }	
  },
  "components": [
    {
      "bom-ref": "pkg:maven/org.acme/common-util@3.0.0",
      "type": "library",
      "group": "org.acme",
      "name": "common-util",
      "version": "3.0.0",
      "purl": "pkg:maven/org.acme/common-util@3.0.0"
    }
  ],
  "dependencies": [
    {
      "ref": "pkg:maven/org.acme/web-framework@1.0.0",
      "dependsOn": [
        "pkg:maven/org.acme/common-util@3.0.0"
      ]
    },
    {
      "ref": "pkg:maven/org.acme/common-util@3.0.0",
      "dependsOn": []
    }
  ]
}

The result of hierarchical merge is: 
{
  "bomFormat": "CycloneDX",
  "specVersion": "1.4",
  "serialNumber": "urn:uuid:3201f6fe-1bd6-4c75-8b15-ffb802fe215c",
  "version": 1,
  "metadata": {
    "component": {
      "type": "application",
      "bom-ref": "Acme@1",
      "name": "Acme",
      "version": "1"
    }
  },
  "components": [
    {
      "type": "library",
      "bom-ref": "org.acme.persistence@3.1.0:pkg:maven/org.acme/persistence@3.1.0",
      "group": "org.acme",
      "name": "persistence",
      "version": "3.1.0",
      "purl": "pkg:maven/org.acme/persistence@3.1.0",
      "components": [
        {
          "type": "library",
          "bom-ref": "org.acme.persistence@3.1.0:pkg:maven/org.acme/common-util@3.0.0",
          "group": "org.acme",
          "name": "common-util",
          "version": "3.0.0",
          "purl": "pkg:maven/org.acme/common-util@3.0.0"
        }
      ]
    },
    {
      "type": "library",
      "bom-ref": "org.acme.web-framework@1.0.0:pkg:maven/org.acme/web-framework@1.0.0",
      "group": "org.acme",
      "name": "web-framework",
      "version": "1.0.0",
      "purl": "pkg:maven/org.acme/web-framework@1.0.0",
      "components": [
        {
          "type": "library",
          "bom-ref": "org.acme.web-framework@1.0.0:pkg:maven/org.acme/common-util@3.0.0",
          "group": "org.acme",
          "name": "common-util",
          "version": "3.0.0",
          "purl": "pkg:maven/org.acme/common-util@3.0.0"
        }
      ]
    }
  ],
  "dependencies": [
    {
      "ref": "org.acme.persistence@3.1.0:pkg:maven/org.acme/persistence@3.1.0",
      "dependsOn": [
        "org.acme.persistence@3.1.0:pkg:maven/org.acme/common-util@3.0.0"
      ]
    },
    {
      "ref": "org.acme.persistence@3.1.0:pkg:maven/org.acme/common-util@3.0.0",
      "dependsOn": []
    },
    {
      "ref": "org.acme.web-framework@1.0.0:pkg:maven/org.acme/web-framework@1.0.0",
      "dependsOn": [
        "org.acme.web-framework@1.0.0:pkg:maven/org.acme/common-util@3.0.0"
      ]
    },
    {
      "ref": "org.acme.web-framework@1.0.0:pkg:maven/org.acme/common-util@3.0.0",
      "dependsOn": []
    },
    {
      "ref": "Acme@1",
      "dependsOn": [
        "org.acme.persistence@3.1.0:pkg:maven/org.acme/persistence@3.1.0",
        "org.acme.web-framework@1.0.0:pkg:maven/org.acme/web-framework@1.0.0"
      ]
    }
  ]
}

The dependency graph of the merge is:
![grafik](https://user-images.githubusercontent.com/79036211/210789446-3627a3fd-ca2c-43be-a07b-42d305a884b8.png)

The expected graph would be:
![grafik](https://user-images.githubusercontent.com/79036211/210789555-cdddd5c5-2f38-46dc-ac8b-742ae1503800.png)

The problem is the following section in the merged file:
  "components": [
    {
      "type": "library",
      "bom-ref": "org.acme.persistence@3.1.0:pkg:maven/org.acme/persistence@3.1.0",
      "group": "org.acme",
      "name": "persistence",
      "version": "3.1.0",
      "purl": "pkg:maven/org.acme/persistence@3.1.0",
      "components": [
        {
          "type": "library",
          "bom-ref": "org.acme.persistence@3.1.0:pkg:maven/org.acme/common-util@3.0.0",
          "group": "org.acme",
          "name": "common-util",
          "version": "3.0.0",
          "purl": "pkg:maven/org.acme/common-util@3.0.0"
        }
The common-util component is described as a kind of sub-component of persistence. If I edit it an put the component
common-util at the same level as persistence ist works fine.

What's the issue: Does CycloneDX CLI merges the wrong way or is the import procedure of DT buggy?

Thanks a lot for your help
Jörg Brünner

**Comment by msymons** [comment]
FYI @coderpatros

**Comment by syalioune** [comment, marked as answer]
From what I can see 

- This is the CLI expected behavior : c.f [related test case](https://github.com/CycloneDX/cyclonedx-dotnet-library/blob/main/tests/CycloneDX.Utils.Tests/MergeTests.cs#L144-L251) and [result sbom](https://github.com/CycloneDX/cyclonedx-dotnet-library/blob/main/tests/CycloneDX.Utils.Tests/__snapshots__/MergeTests.HierarchicalMergeComponentsTest.snap)
- Only first level components are considered by DT when parsing dependencies. Nested components are ignored.
